sympy/solvers/ode.py

```diff
@@ -1198,8 +1198,12 @@ class in it.  Note that a hint may do this anyway if
 
         # This match is used for several cases below; we now collect on
         # f(x) so the matching works.
+        roots = solve(reduced_eq,df)
+        r1 = collect(df-roots[0], df, exact=True).match(d + e*df)
         r = collect(reduced_eq, df, exact=True).match(d + e*df)
```

Comment:
If `r` succeeds then you won't need `r1` so how about delay until you need it:
```python
r = collect(reduced_eq, df, exact=True).match(d + e*df)
if r is None:
    roots = solve(reduced_eq, df)
    if roots:
        root = roots[0]  # assuming all will give the same result
        r = {d: root, e: S.One}
if r:
    # Using r[d] ...
```
And if you solved for `df`, is collection and matching necessary? My suggestion above assumes that it is not.

Reply:
yes we can use this as i was just sharing my approach i tried that code.But what we we should do when we have different roots/

Reply:
That would be better. We do need to consider all roots though because they don't all give the same solutions e.g. (with this PR):
```julia
In [21]: eq = f(x).diff(x)**2 - f(x).diff(x)                                                                                      

In [22]: eq                                                                                                                       
Out[22]: 
          2           
⎛d       ⎞    d       
⎜──(f(x))⎟  - ──(f(x))
⎝dx      ⎠    dx      

In [23]: dsolve(eq, hint='nth_algebraic')                                                                                         
Out[23]: [f(x) = C₁, f(x) = C₁ + x]

In [24]: dsolve(eq, hint='lie_group')                                                                                             
Out[24]: [f(x) = C₁, f(x) = C₁ + x]

In [25]: dsolve(eq, hint='separable')                                                                                             
Out[25]: f(x) = C₁
```
I suggest maybe:
```python
r = collect(reduced_eq, df, exact=True).match(d + e*df)
if r is None:
    roots = solve(reduced_eq, df)
    if roots:
        r_all = [{d: root, e: S.One} for root in roots]
        r = r_all[0]
if r:
    # Use r_all with the separable method but still use r for the other methods.
```
Then here:
https://github.com/sympy/sympy/blob/2c3d7dd26f256b4df37df6ad29bb8571715965c9/sympy/solvers/ode.py#L1224-L1225
We need to store a list of solutions rather than a single one.

Reply:
This code block is behaving unexpectedly as it goes in infinite recursion

Reply:
What about:
```python
        if r is None:
            roots = solve(reduced_eq, df)
            if roots:
                meq = Mul(*[(df - i) for i in roots])
                m = _ode_factorable_match(meq, func, kwargs.get('x0', 0))
                matching_hints['factorable'] = m
```

Reply:
`classify_ode(eq)` will give `factorable` as a hint instead of `separable` i guess

Reply:
That seems reasonable. I guess that's more or less equivalent to trying to solve for `f(x).diff(x)` right back at the beginning of `dsolve` which is something I've thought about a few times. Maybe it should check if the `factorable` hint has already matched before calling `solve`.

Reply:
`dsolve(f(x).diff(x)**2 - f(x)**3)`
returns `Eq(f(x), 4/(C1**2 - 2*C1*x + x**2))` after making changes as suggested above

Reply:
so, should i commit the suggested code block?

Reply:
> That seems reasonable. I guess that's more or less equivalent to trying to solve for `f(x).diff(x)` right back at the beginning of `dsolve` which is something I've thought about a few times. 

I think we should try to avoid using `solve` and let pattern matching be the priority and only fall back to `solve` if no pattern has matched.

> Maybe it should check if the `factorable` hint has already matched before calling `solve`.

That's a good recommendation, too.

Reply:
![Screenshot from 2020-01-04 22-54-04](https://user-images.githubusercontent.com/44258119/71769307-6ad7d500-2f45-11ea-9e44-41800b1df981.png)
does this look good?

Reply:
It's easier to comment on the change if you push it to the PR.

We mustn't use bare except anywhere in SymPy though. Probably better to do
```python
if r is None and 'factorable' not in matching_hints:
```

Reply:
Should we delay this solve attempt until the end of the routine when we find that no hint has succeeded?

Reply:
> Should we delay this solve attempt until the end of the routine when we find that no hint has succeeded?

I wouldn't bother with that. This whole routine is incredibly inefficient because it tries to match everything before returning anything. A refactor could make this much faster but in the mean-time we shouldn't overcomplicate as that will make the refactor harder.